Prioritise the legacy VMM address during initialisation

Xeeynamo opened the pull request:
This pull request aims to fix the problem described in #3638  .

The new system introduced in #3523 proposes various base addresses and check if they can be actually used by the Virtual Memory Manager. This change just proposes to try to initialise the VMM by using `0x20000000` as base address first, for ~both x64 and~ x86 builds.
Of course this does not guarantee that the legacy memory address will be used, but it will help third party software to work to new builds and to slowly migrate to an API that eventually will be exposed by the emulator.

tellowkrinkle commented:
On Ubuntu 18.04.5 x64:
Assertion Failed: SIB target is too far away, needs an indirect register
stack trace:
![stack trace screenshot](https://cdn.discordapp.com/attachments/518553380264607786/749713985854504971/unknown.png)

tellowkrinkle commented on pcsx2/System.cpp:
```diff
@@ -358,6 +358,18 @@ namespace HostMemoryMap {
 /// Attempts to find a spot near static variables for the main memory
 static VirtualMemoryManagerPtr makeMainMemoryManager() {
+	const uptr UpperBounds = 0U;
+	const bool Strict = true;
```
I'd prefer if the fact that the value was `true` was clearly visible at the call site, not way at the top of the function.  This option is decided separately for each time we try to make a VMM, and isn't even the same for all three.

tellowkrinkle replied:
If you want to consolidate things, make a
```cpp
auto tryMakeAt = [&memoryManagerName](uptr base) -> VirtualMemoryManagerPtr {
	std::make_shared<VirtualMemoryManager>(memoryManagerName, base, HostMemoryMap::Size, /*upper_bounds=*/0, /*strict=*/true);
	if (mgr->IsOk())
 		return mgr;
	return nullptr;
}
```
Then you can
```cpp
if (auto mgr = tryMakeAt(LegacyMemoryBase))
	return mgr;
```

Xeeynamo commented:
Hi @tellowkrinkle and thanks for the code review, I modified the source code as you suggested.

Since it is just refactoring, I expect that the assertion will still continue to fail. Could you provide me more details on how to reproduce this crash? I built and ran it on both Windows 10 x64 and Arch Linux with Multilib and tested with SLUS-21174. I had no problems with `master` and the change I performed.

I rebased this branch to the latest changed on `master` just in case.

Please keep me updated 🙂

tellowkrinkle commented:
It's caused by the `0x20000000` allocation being too far away from `SSEXYZWMask[0]` referenced in `newVif_UnpackSSE.cpp:188`

Reproducible by attempting to boot the bios

Xeeynamo commented:
@tellowkrinkle I was able to reproduce the crash and it is x64 specific. Therefore I added a condition to not attempt to use that "magic" address on x64 builds.

Please let me know if there are other concerns or if it is ready to be approved.

tellowkrinkle commented on pcsx2/System.cpp:
```diff
@@ -358,6 +358,27 @@ namespace HostMemoryMap {
+	// tools relies to this specific address, therefore try to initialize the memory manager using this base
+	// address first. This does not guarantee it, but it prioritise the legacy memory address.
+	// This is only valid for x86 builds, as x64 have a high chance of failing an assertion by using this
+	// specific address. x64 builds currently uses 0x60000000.
```
Position in x64 depends on where your linker decides to put PCSX2, and will vary by platform, so I wouldn't put it in the comments here

e.g. on my computer (macOS 10.15) it's at 0x140000000

Xeeynamo replied:
Comment removed!

tellowkrinkle commented on pcsx2/System.cpp:
```diff
@@ -358,6 +358,27 @@ namespace HostMemoryMap {
+	// address first. This does not guarantee it, but it prioritise the legacy memory address.
+	// This is only valid for x86 builds, as x64 have a high chance of failing an assertion by using this
+	// specific address.
+	//if (sizeof(void*) == 4)
```
You accidentally left the `//` in

Xeeynamo replied:
That was definitively a distraction. Thanks for the flawless code review.

tellowkrinkle reviewed (approved):
BTW just so you know, 32-bit support isn't going to stay forever, though I'm happy to keep the old memory layout as long as we continue support it

Xeeynamo commented:
> BTW just so you know, 32-bit support isn't going to stay forever, though I'm happy to keep the old memory layout as long as we continue support it

Yes, I imagined that 😃. But as the description of the PR says, it would be good to give a window of migration for the existing 3rd party programs to a more future-proof solution (eg. the IPC PR).
Thanks for your time to review this!

F0bes commented:
Closing as the next release will be 64-bit only and this will have no effect.
